src/backend/distributed/commands/utility_hook.c

```diff
@@ -468,6 +468,36 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
 		PreprocessTruncateStatement((TruncateStmt *) parsetree);
 	}
 
+	if (IsA(parsetree, LockStmt))
```

Comment:
We are in the pre-process. Postgres does the following checks during StandardProcess utility. Should we do the same? Otherwise, underprivileged users would be able to block tables briefly until `StandardProcess` does its checks?


```
LOCK TABLE ... IN ACCESS SHARE MODE requires SELECT privileges on the target table. LOCK TABLE ... IN ROW EXCLUSIVE MODE requires INSERT, UPDATE, DELETE, or TRUNCATE privileges on the target table. All other forms of LOCK require table-level UPDATE, DELETE, or TRUNCATE privileges.
```

Reply:
actually, never mind, we'd get error from the worker. 

But still we'd be unnecessarily establishing connections in that case. Can we check privileges before sending the command? But we should be careful to release the locks

Reply:
Preprocess happens before standard utility, but its DDL tasks are executed after standard utility.

We should not acquire locks locally before acquiring remote lock on the first node. Otherwise, concurrent `LOCK TABLE` commands from different workers would cause a distributed deadlock. 

Is that your question?

Reply:
I don't follow. Why can't we call AcquireDistributedLockOnRelations in preprocess?

Reply:
(What would be even nicer is for distributed_object_ops to add an optional lockLevel function)

Reply:
To touch on some of the things discussed here and share my thoughts on how I intend to implement stuff:
* Permission checks: I can check for permissions before a command is sent to the worker to avoid creating an unnecessary connection. In cases when there is no permission issue (which I suspect would be most cases), we are doing an extra redundant check which should be cheap so it shouldn't be a problem.
* How to distribute the command: I can create a LockStmt distributed_object_ops which does exactly what the current implementation does, but it does it in a preprocess function. Afaict, this only contributes to code uniformity and has no other implications. On that note, we might consider doing the same thing for TRUNCATE (since my current approach was pretty much using the same logic TRUNCATE is using).

Reply:
I think handling the locking in TruncateStmt and LockStmt in a similar way would be great.

Reply:
> I don't follow. Why can't we call AcquireDistributedLockOnRelations in preprocess?
> What could be even nicer is a dedicated locking function in distributed_object_ops.

A typical `.preprocess` function happens [before](https://github.com/citusdata/citus/blob/6254f30305df3201b1f4da8ed8161705f4b00c1a/src/backend/distributed/commands/utility_hook.c#L491-L494)  [standardProcessUtility](https://github.com/citusdata/citus/blob/6254f30305df3201b1f4da8ed8161705f4b00c1a/src/backend/distributed/commands/utility_hook.c#L594). This is good. But, again a typical `.preprocess` only creates the [`DDL` tasks](https://github.com/citusdata/citus/blob/6254f30305df3201b1f4da8ed8161705f4b00c1a/src/backend/distributed/commands/statistics.c#L100). And, the tasks are [executed](https://github.com/citusdata/citus/blob/6254f30305df3201b1f4da8ed8161705f4b00c1a/src/backend/distributed/commands/utility_hook.c#L682-L685) only after `standardProcessUtility` happens.

So, assume the following scenario:

1) `Session-1` on `Worker-1` does `LOCK TABLE test ACCESS EXCLUSIVE` 
2) Command on (1) is executed locally on `Worker 1`
3) `Session-2` on `Worker-2` does `LOCK TABLE test ACCESS EXCLUSIVE` 
4) Command on (3) is executed locally on `Worker 2`
5) Command on (1) is blocked on `Worker -2` behind command (3)
6) Command on (3) is blocked on `Worker -1` behind command (1)
7) Distributed deadlock happens that would not happen on local PG

Reply:
@onderkalaci 
Wouldn't 
```c
List * PreprocessLockStmt(...){
	// extract the relations
	AcquireDistributedLockOnRelations(...);
	return NIL;
}
```
this work. No DDL tasks created, only calling `AcquireDistributedLockOnRelations` before the `standardProcessUtility` so there will be no deadlocks like the scenario you explained.

Reply:
Yes, could work. 

but then it feels a little unusual to acquire the locks in `PreprocessLockStmt `, because we don't do such things in `PreprocessXXX` functions. Instead, it feels like doing it here explicitly easier to follow (at least for me)

Reply:
> This is good. But, again a typical .preprocess only creates the [DDL tasks](https://github.com/citusdata/citus/blob/6254f30305df3201b1f4da8ed8161705f4b00c1a/src/backend/distributed/commands/statistics.c#L100).

If this is the case then should we rename preprocess to something like createDDLTaskList to make that more explicit?

> Instead, it feels like doing it here explicitly easier to follow (at least for me)

This is a giant function with a lot of statement-specific logic already. We're doing a similar thing for TruncateStmt and LockStmt, and may want to do this for other statements in the future. We should come up with a more generic solution.

Reply:
> If this is the case then should we rename preprocess to something like createDDLTaskList to make that more explicit?

We also sometimes create DDL tasks on PostprocessXXX. So, maybe `PreProcessCreateDDLTaskList` and `PostProcessCreateDDLTaskList` could work?

> more generic solution.

So, one idea you noted above could make it more generic, like adding `DistributeObjectOps->DistributedSetup` function or such could make sense

Reply:
Afaict most actions currently done explicitly in `ProcessUtilityInternal` could be done in such `DistributeObjectOps->DistributedSetup`.
There is also some post process logic which does not return ddl task. So we might want to add `DistributeObjectOps->DistributedPostProcess`.

There is also the issue of `EnableMetadataSync` being on or off.
Some logic is executed always (no matter if it is on or off) while some only when the flag is turned on. A generic solution would ideally also address this distinction.

Reply:
> So, maybe PreProcessCreateDDLTaskList and PostProcessCreateDDLTaskList could work?
DDL tasks must be returned exclusively by either the preprocess or the postprocess. Maybe having two functions for this is redundant (though I need to go over the use cases of this function to make sure this is indeed the case) and we can have
* Preprocess
* CreateDDLTaskList
* Postprocess

Where preprocess and postprocess don't return anything.

Reply:
>  Maybe having two functions for this is redundant


In the code, we sometime create DDL tasks before standardProcessUtility and sometimes after. Not sure if that is strictly necessary, but most likely it is a large refactor that we wouldn't want in 11.

For this patch, we should probably at most consider adding `DistributedSetup `, and then do the other refactors on master branch

Reply:
Thinking a bit more, do we need to do QualifyTreeNode before doing the locking to ensure we use the same schema in locking?

Maybe we should have a call on this (though would be Tuesday at the earliest)

Reply:
> Maybe we should have a call on this (though would be Tuesday at the earliest)

Sure, I'll see what I can figure out independently until then.
Plus I can still add the tests which should be mostly independent of the implementation.

Reply:
would be nice to add a comment why we do this operation before anything done (e.g., before locally acquiring the lock